File: 0986-interval-list-intersections/0986-interval-list-intersections.py
class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        ans = []
        i = j = 0

        while i < len(firstList) and j < len(secondList):
            si, ei = firstList[i]
            sj, ej = secondList[j]

            start = max(si, sj)
            end = min(ei, ej)

            if start <= end:
                ans.append([start, end])

            if ei < ej:
                i += 1
            else:
                j += 1

        return ans

File: 0986-interval-list-intersections/0986-interval-list-intersections.cpp
class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& firstList, vector<vector<int>>& secondList) {
        vector<vector<int>> ans;
        int n = firstList.size();
        int m = secondList.size();

        int i = 0, j = 0;

        while(i < n && j < m){
            int si = firstList[i][0];
            int ei = firstList[i][1];
            int sj = secondList[j][0];
            int ej = secondList[j][1];

            int st = max(si, sj);
            int end = min(ei, ej);

            if(st <= end) ans.push_back({st, end});

            if(ei < ej) i++;
            else j++;
        }

        return ans;
    }
};